File: Graphs/P024-025.py
from itertools import permutations
def is_valid_order(order, edges, pos):
    for u, v in edges:
        if pos[v] > pos[u]: 
            return False
    return True

def Brute(N,P,array):
    vertices = list(range(N))
    for perm in permutations(vertices):
        pos = {perm[i]: i for i in range(N)}

        if is_valid_order(perm, array, pos):
            return True 
    return False

from collections import deque
def optimal(N,P,array):
    indegree=[0]*N
    adj=[[] for _ in range(N)]
    for u,v in array:
        adj[v].append(u)
        indegree[u]+=1
    q=deque()
    for i in range(N):
        if indegree[i]==0:
            q.append(i)
    res=[]
    count=0
    while q:
        node=q.popleft()
        count+=1
        res.append(node)
        for nei in adj[node]:
            indegree[nei]-=1
            if indegree[nei]==0:
                q.append(nei)
    return count==N

if __name__=='__main__':
    N = 4
    P = 3
    array=[(1,0),(2,1),(3,2)]
    print("Brute: ",Brute(N,P,array))
    print("Optimal: ",optimal(N,P,array))

File: Striver_C++_Problems/Graphs/P024-025.cpp
// P024_025.cpp - Check if all tasks can be finished

#include <bits/stdc++.h>
using namespace std;

bool is_valid_order2(const vector<int>& order,
                     const vector<pair<int,int>>& edges,
                     const vector<int>& pos) {
    for (auto &e : edges) {
        int u = e.first;
        int v = e.second;
        // in Python: if pos[v] > pos[u]: return False
        // so requirement is pos[v] <= pos[u]
        if (pos[v] > pos[u]) return false;
    }
    return true;
}

// Brute: try all permutations, return True if any respects constraints
bool Brute(int N, int P, const vector<pair<int,int>>& array) {
    vector<int> vertices(N);
    iota(vertices.begin(), vertices.end(), 0);
    sort(vertices.begin(), vertices.end());

    do {
        vector<int> pos(N);
        for (int i = 0; i < N; i++) pos[vertices[i]] = i;
        if (is_valid_order2(vertices, array, pos)) return true;
    } while (next_permutation(vertices.begin(), vertices.end()));

    return false;
}

// Optimal: Kahn, using the same graph they used (edge v->u)
bool optimal(int N, int P, const vector<pair<int,int>>& array) {
    vector<vector<int>> adj(N);
    vector<int> indegree(N, 0);
    for (auto &e : array) {
        int u = e.first, v = e.second;
        adj[v].push_back(u);
        indegree[u]++;
    }
    queue<int> q;
    for (int i = 0; i < N; i++)
        if (indegree[i] == 0) q.push(i);

    int count = 0;
    vector<int> res;
    while (!q.empty()) {
        int node = q.front(); q.pop();
        count++;
        res.push_back(node);
        for (int nei : adj[node]) {
            indegree[nei]--;
            if (indegree[nei] == 0) q.push(nei);
        }
    }
    return count == N;
}

int main() {
    int N = 4;
    int P = 3;
    vector<pair<int,int>> array = {{1,0},{2,1},{3,2}};

    cout << "Brute: "  << (Brute(N, P, array) ? "True" : "False") << "\n";
    cout << "Optimal: " << (optimal(N, P, array) ? "True" : "False") << "\n";
    return 0;
}
